aio.api.github/aio/api/github/abstract/repo.py

```diff
@@ -0,0 +1,133 @@
+
+import pathlib
+from datetime import datetime
+from functools import cached_property, partial
+from typing import Any, Dict, Optional, Type
+
+import abstracts
+
+from aio.api.github import abstract, utils
+from . import base
+
+
+class AGithubRepo(metaclass=abstracts.Abstraction):
+    """A Github repo."""
+
+    def __init__(self, github: "abstract.AGithubAPI", name: str) -> None:
+        self.github = github
+        self.name = name
+
+    @cached_property
+    def github_path(self) -> pathlib.PurePosixPath:
+        """Github API repos path."""
+        return pathlib.PurePosixPath(f"/repos/{self.name}")
+
+    @cached_property
+    def issues(self) -> "abstract.AGithubIssues":
+        """Github issues for this repo."""
+        return self.github.issues_class(
+            self.github,
+            repo=self)
+
+    @property
+    def labels(self) -> "abstract.AGithubIterator":
+        """Github labels for this repo."""
+        return self.iter_entities(
+            self.github.label_class, "labels")
+
+    async def commit(self, name: str) -> "abstract.AGithubCommit":
+        """Fetch a commit for this repo."""
+        return self.github.commit_class(
+            self,
+            await self.getitem(f"commits/{name}"))
+
+    def commits(self, since: datetime = None) -> "abstract.AGithubIterator":
+        """Iterate commits for this repo."""
+        query = "commits"
+        if since is not None:
+            query = f"{query}?since={utils.dt_to_js_isoformat(since)}"
+        return self.getiter(
+            query,
+            inflate=partial(self.github.commit_class, self))
+
+    async def getitem(self, query: str) -> Any:
+        """Call the `gidgethub.getitem` api for this repo."""
+        return await self.github.getitem(self.github_endpoint(query))
+
+    def getiter(self, query: str, **kwargs) -> "abstract.AGithubIterator":
+        """Return a `GithubIterator` wrapping `gidgethub.getiter` for this
+        repo."""
+        return self.github.getiter(self.github_endpoint(query), **kwargs)
+
+    def github_endpoint(self, rel_path: str) -> str:
+        """Github API path for provided relative path."""
+        return str(self.github_path.joinpath(rel_path))
+
+    def iter_entities(
+            self,
+            entity: Type[base.GithubEntity],
+            path: str,
+            **kwargs) -> "abstract.AGithubIterator":
+        """Iterate and inflate entities for provided type."""
+        return self.getiter(
+            path,
+            inflate=partial(entity, self),
+            **kwargs)
+
+    async def newer_release(
```

Comment:
perhaps rename this to `highest_release` - ive struggled to think of a name for this